File: leetcode/73.cpp
// 机智地使用了第一行和第一列用来存储信息，然后 hard coding 手动处理第一行第一列。
class Solution {
public:
    void setZeroes(vector<vector<int>>& v) {
        int m = v.size();
        int n = v[0].size();
        
        bool colzero = false;
        for (int i = 0; i < m; i++) {
            if (v[i][0] == 0)
                colzero = true;
        }
        bool rowzero = false;
        for (int i = 0; i < n; i++) {
            if (v[0][i] == 0)
                rowzero = true;
        }
        
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (v[i][j] == 0) {
                    v[i][0] = 0;
                    v[0][j] = 0;
                }
            }
        }
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (v[i][0] == 0 || v[0][j] == 0) {
                    v[i][j] = 0;
                }
            }
        }
        
        if (rowzero) {
            for (int i = 0; i < n; i++) {
                v[0][i] = 0;
            }
        }
        if (colzero) {
            for (int i = 0; i < m; i++) {
                v[i][0] = 0;
            }
        }
    }
};

File: leetcode/73.rs
impl Solution {
    pub fn set_zeroes(mat: &mut Vec<Vec<i32>>) {
        let m = mat.len();
        if m == 0 {
            return;
        }
        let n = mat[0].len();
        if n == 0 {
            return;
        }

        let mut zrow = None;
        'l: for i in 0..m {
            for j in 0..n {
                if mat[i][j] == 0 {
                    zrow = Some(i);
                    break 'l;
                }
            }
        }
        if zrow.is_none() {
            return;
        }
        let zrow: usize = zrow.unwrap();

        for i in 0..m {
            if i == zrow {
                continue;
            }
            let mut z = false;
            for j in 0..n {
                if mat[i][j] == 0 {
                    z = true;
                    mat[zrow][j] = 0;
                }
            }
            if z {
                for j in 0..n {
                    mat[i][j] = 0;
                }
            }
        }

        for j in 0..n {
            if mat[zrow][j] != 0 {
                mat[zrow][j] = 0;
            } else {
                for i in 0..m {
                    mat[i][j] = 0;
                }
            }
        }
    }
}
